File: Example/CoreData/SyncKit/QSCompanyTableViewController.m
//
//  QSCompanyTableViewController.m
//  SyncKit
//
//  Created by Manuel Entrena on 31/07/2016.
//  Copyright © 2016 Manuel. All rights reserved.
//

#import "QSCompanyTableViewController.h"
#import "QSCompany.h"
#import "QSEmployeeTableViewController.h"
#import "QSCompanyTableViewCell.h"
@import SyncKit;

@interface QSCompanyTableViewController () <NSFetchedResultsControllerDelegate, UICloudSharingControllerDelegate>

@property (nonatomic, strong) NSFetchedResultsController *fetchedResultsController;

@property (nonatomic, weak) IBOutlet UIButton *syncButton;
@property (nonatomic, weak) IBOutlet UIActivityIndicatorView *indicatorView;

@property (nonatomic, strong) QSCompany *sharingCompany;

@property (nonatomic, weak) IBOutlet UIView *loadingView;

@end

@implementation QSCompanyTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupFetchedResultsController];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self.tableView reloadData];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)setupFetchedResultsController
{
    if (!_fetchedResultsController) {
        NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] initWithEntityName:@"QSCompany"];
        fetchRequest.sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:@"name" ascending:YES]];
        
        _fetchedResultsController = [[NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest managedObjectContext:self.managedObjectContext sectionNameKeyPath:nil cacheName:nil];
        _fetchedResultsController.delegate = self;
        [_fetchedResultsController performFetch:nil];
    }
}

- (void)createCompanyWithName:(NSString *)name
{
    QSCompany *company = [NSEntityDescription insertNewObjectForEntityForName:@"QSCompany" inManagedObjectContext:self.managedObjectContext];
    company.identifier = [[NSUUID UUID] UUIDString];
    company.name = name;
    [self.managedObjectContext save:nil];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"showEmployees"]) {
        QSEmployeeTableViewController *employeeTableViewController = (QSEmployeeTableViewController *)segue.destinationViewController;
        employeeTableViewController.managedObjectContext = self.managedObjectContext;
        employeeTableViewController.company = (QSCompany *)sender;
        employeeTableViewController.canWrite = YES;
    }
}

#pragma mark - NSFetchedResultsControllerDelegate

- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller {
    [self.tableView beginUpdates];
}


- (void)controller:(NSFetchedResultsController *)controller didChangeSection:(id <NSFetchedResultsSectionInfo>)sectionInfo
           atIndex:(NSUInteger)sectionIndex forChangeType:(NSFetchedResultsChangeType)type {
    
    switch(type) {
        case NSFetchedResultsChangeInsert:
            [self.tableView insertSections:[NSIndexSet indexSetWithIndex:sectionIndex]
                          withRowAnimation:UITableViewRowAnimationFade];
            break;
            
        case NSFetchedResultsChangeDelete:
            [self.tableView deleteSections:[NSIndexSet indexSetWithIndex:sectionIndex]
                          withRowAnimation:UITableViewRowAnimationFade];
            break;
        default:
            break;
    }
}


- (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)anObject
       atIndexPath:(NSIndexPath *)indexPath forChangeType:(NSFetchedResultsChangeType)type
      newIndexPath:(NSIndexPath *)newIndexPath {
    
    UITableView *tableView = self.tableView;
    
    switch(type) {
            
        case NSFetchedResultsChangeInsert:
            [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
            
        case NSFetchedResultsChangeDelete:
            [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
            
        case NSFetchedResultsChangeUpdate:
            [self configureCell:[tableView cellForRowAtIndexPath:indexPath]
                    atIndexPath:indexPath];
            break;
            
        case NSFetchedResultsChangeMove:
            [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
    }
}


- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller {
    [self.tableView endUpdates];
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewCellEditingStyleDelete;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    QSCompany *company = [self.fetchedResultsController objectAtIndexPath:indexPath];
    [self performSegueWithIdentifier:@"showEmployees" sender:company];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return [[self.fetchedResultsController sections] count];
}

- (NSInteger)tableView:(UITableView *)table numberOfRowsInSection:(NSInteger)section {
    if ([[self.fetchedResultsController sections] count] > 0) {
        id <NSFetchedResultsSectionInfo> sectionInfo = [[self.fetchedResultsController sections] objectAtIndex:section];
        return [sectionInfo numberOfObjects];
    } else
        return 0;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    QSCompanyTableViewCell *cell = (QSCompanyTableViewCell *)[tableView dequeueReusableCellWithIdentifier:@"cell"];
    [self configureCell:cell atIndexPath:indexPath];

    return cell;
}

- (void)configureCell:(QSCompanyTableViewCell *)cell atIndexPath:(NSIndexPath *)indexPath
{
    QSCompany *company = [self.fetchedResultsController objectAtIndexPath:indexPath];
    cell.nameLabel.text = company.name;
    CKShare *share = [self.synchronizer shareFor:company];
    if (share) {
        [cell.sharingButton setTitle:@"Sharing" forState:UIControlStateNormal];
    } else {
        [cell.sharingButton setTitle:@"Share" forState:UIControlStateNormal];
    }
    
    __weak QSCompanyTableViewController *weakSelf = self;
    
    cell.shareButtonAction = ^{
        [weakSelf shareCompany:company];
    };
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    return YES;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        QSCompany *company = [self.fetchedResultsController objectAtIndexPath:indexPath];
        [self.managedObjectContext deleteObject:company];
        [self.managedObjectContext save:nil];
    }
}

#pragma mark - Actions

- (IBAction)insertNewCompany:(id)sender
{
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"New company" message:nil preferredStyle:UIAlertControllerStyleAlert];
    [alertController addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = @"Enter company's name";
    }];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Add" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self createCompanyWithName:alertController.textFields.firstObject.text];
    }]];
    [self presentViewController:alertController animated:YES completion:nil];
}

- (IBAction)didTapSynchronize:(id)sender
{
    [self synchronizeWithCompletion:nil];
}

- (void)synchronizeWithCompletion:(void(^)(NSError *error))completion
{
    [self showLoading:YES];
    __weak QSCompanyTableViewController *weakSelf = self;
    [self.synchronizer synchronizeWithCompletion:^(NSError *error) {
        
        [weakSelf showLoading:NO];
        
        if (error) {
            if (error.code == CKErrorChangeTokenExpired) {
                [self.appDelegate didGetChangeTokenExpiredError];
            } else {
                UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Error" message:[NSString stringWithFormat:@"Error: %@", error] preferredStyle:UIAlertControllerStyleAlert];
                
                [alertController addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
                
                [weakSelf presentViewController:alertController animated:YES completion:nil];
            }
        } else {
            CKRecordZoneID *zoneID = weakSelf.synchronizer.modelAdapters.firstObject.recordZoneID;
            if (zoneID) {
                [weakSelf.synchronizer subscribeForChangesIn:zoneID completion:^(NSError *error) {
                    if (error) {
                        NSLog(@"Failed to subscribe with error: %@", error);
                    } else {
                        NSLog(@"Subscribed for notifications");
                    }
                }];
            }
        }
        if (completion) {
            completion(error);
        }
    }];
}

- (void)shareCompany:(QSCompany *)company
{
    self.sharingCompany = company;
    
    __weak QSCompanyTableViewController *weakSelf = self;
    
    [self synchronizeWithCompletion:^(NSError *error) {
        
        if (!error) {
            UICloudSharingController *sharingController;
            CKShare *share = [weakSelf.synchronizer shareFor:company];
            CKContainer *container = [CKContainer containerWithIdentifier:weakSelf.synchronizer.containerIdentifier];
            if (share) {
                share[CKShareTitleKey] = company.name;
                sharingController = [[UICloudSharingController alloc] initWithShare:share container:container];
            } else {
                sharingController = [[UICloudSharingController alloc] initWithPreparationHandler:^(UICloudSharingController * _Nonnull controller, void (^ _Nonnull preparationCompletionHandler)(CKShare * _Nullable, CKContainer * _Nullable, NSError * _Nullable)) {
                    
                    [self.synchronizer shareWithObject:company publicPermission:CKShareParticipantPermissionReadOnly participants:@[] completion:^(CKShare *share, NSError *error) {
                        
                        share[CKShareTitleKey] = company.name;
                        preparationCompletionHandler(share, container, error);
                    }];
                }];
            }
            
            sharingController.availablePermissions = UICloudSharingPermissionAllowPublic | UICloudSharingPermissionAllowReadOnly | UICloudSharingPermissionAllowReadWrite;
            sharingController.delegate = self;
            [weakSelf presentViewController:sharingController animated:YES completion:nil];
        }
    }];
}

#pragma mark - UICloudSharingControllerDelegate

- (NSString *)itemTitleForCloudSharingController:(UICloudSharingController *)csc
{
    return self.sharingCompany.name;
}

- (void)cloudSharingController:(UICloudSharingController *)csc failedToSaveShareWithError:(NSError *)error
{
    NSLog(@"Error saving: %@", error);
}

- (void)cloudSharingControllerDidSaveShare:(UICloudSharingController *)csc
{
    [self.synchronizer saveShare:csc.share for:self.sharingCompany];
    [self.tableView reloadData];
}

- (void)cloudSharingControllerDidStopSharing:(UICloudSharingController *)csc
{
    [self.synchronizer deleteShareFor:self.sharingCompany];
    [self.tableView reloadData];
}

#pragma mark - Loading

- (void)showLoading:(BOOL)loading
{
    if (loading) {
        self.syncButton.hidden = YES;
        self.loadingView.hidden = NO;
        [self.indicatorView startAnimating];
    } else {
        self.syncButton.hidden = NO;
        self.loadingView.hidden = YES;
        [self.indicatorView stopAnimating];
    }
}

@end

File: Example/CoreData/SyncKit/QSCompanyTableViewController.swift
//
//  QSCompanyTableViewController.swift
//  SyncKitCoreDataExample
//
//  Created by Jérôme Haegeli on 24.07.18.
//  Copyright © 2018 Manuel. All rights reserved.
//

import UIKit
import SyncKit

class QSCompanyTableViewController: UITableViewController, NSFetchedResultsControllerDelegate, UICloudSharingControllerDelegate {
    
    var managedObjectContext: NSManagedObjectContext?
    var synchronizer: QSCloudKitSynchronizer?
    
    var fetchedResultsController: NSFetchedResultsController<NSFetchRequestResult>?
    @IBOutlet weak var syncButton: UIButton!
    @IBOutlet weak var indicatorView: UIActivityIndicatorView!
    
    var sharingCompany: QSCompany!
    
    @IBOutlet weak var loadingView: UIView?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        managedObjectContext = CoreDataStack.shared.persistentContainer.viewContext
        synchronizer = CoreDataStack.shared.synchronizer
        setupFetchedResultsController()
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(true)
        self.tableView.reloadData()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    func setupFetchedResultsController() {
        if !(fetchedResultsController != nil) {
            let fetchRequest = NSFetchRequest<NSFetchRequestResult>(entityName: "QSCompany")
            fetchRequest.sortDescriptors = [NSSortDescriptor(key: "name", ascending: true)]
            fetchedResultsController = NSFetchedResultsController(fetchRequest: fetchRequest, managedObjectContext: managedObjectContext!, sectionNameKeyPath: nil, cacheName: nil)
            fetchedResultsController?.delegate = self
            try? fetchedResultsController?.performFetch()
            let array = try? managedObjectContext?.fetch(fetchRequest)
        }
    }


    func createCompanyWithName(name: String) {
        
        let company = QSCompany(context: self.managedObjectContext!)
            company.identifier = UUID().uuidString
            company.name = name
            do {
                try self.managedObjectContext?.save()
            } catch {
                print(error)
            }
        
     }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "showEmployees" {
           
            guard let employeeTableViewController = segue.destination as? QSEmployeeTableViewController else {
                    fatalError("Application storyboard mis-configuration")
                }
            
            let company = sender as? QSCompany
            employeeTableViewController.company = company
            employeeTableViewController.managedObjectContext = self.managedObjectContext
            employeeTableViewController.canWrite = true
        }
    }
    
//pragma mark - NSFetchedResultsControllerDelegate
  
    
     func controllerWillChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        self.tableView.beginUpdates()
    }
    
    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange sectionInfo: NSFetchedResultsSectionInfo, atSectionIndex sectionIndex: Int, for type: NSFetchedResultsChangeType) {
        switch type {
        case .insert:
            tableView.insertSections(NSIndexSet(index: sectionIndex) as IndexSet, with: .fade)
        case .delete:
            tableView.deleteSections(NSIndexSet(index: sectionIndex) as IndexSet, with: .fade)
        default:
            break
        }
    }

    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange anObject: Any, at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?) {
        let tableView: UITableView? = self.tableView
        switch type {
        case .insert:
            tableView?.insertRows(at: [newIndexPath!], with: .fade)
        case .delete:
            tableView?.deleteRows(at: [indexPath!], with: .fade)
        case .update:
            if let aPath = indexPath {
                configureCell(cell: tableView?.cellForRow(at: aPath) as! QSCompanySwiftTableViewCell, atIndexPath: indexPath!)
            }
        case .move:
            tableView?.deleteRows(at: [indexPath!], with: .fade)
            tableView?.insertRows(at: [newIndexPath!], with: .fade)
        }
    }

    func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        tableView.endUpdates()
    }

    override func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCellEditingStyle {
        return .delete
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let company = fetchedResultsController?.object(at: indexPath) as? QSCompany
        performSegue(withIdentifier: "showEmployees", sender: company)
    }

    
    
    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        if let frc = fetchedResultsController {
            return frc.sections!.count
        }
        return 0
    }

    override func tableView(_ table: UITableView, numberOfRowsInSection section: Int) -> Int {
        if (fetchedResultsController?.sections?.count)! > 0 {
            let sectionInfo = fetchedResultsController?.sections![section]
            return sectionInfo!.numberOfObjects
        } else {
            return 0
        }
    }

   
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath) as! QSCompanySwiftTableViewCell

        // Configure the cell...

        configureCell(cell: cell, atIndexPath: indexPath)
        return cell
    }
    
    func configureCell(cell: QSCompanySwiftTableViewCell, atIndexPath indexPath: IndexPath) {
        
        let company = self.fetchedResultsController?.object(at: indexPath) as! QSCompany
        cell.nameLabel.text = company.name
        let share = self.synchronizer?.share(for: company)
        if (share != nil) {
            cell.sharingButton.setTitle("Sharing", for: .normal)
        } else {
            cell.sharingButton.setTitle("Share", for: .normal)
        }
        
        cell.shareButtonAction = {
            self.shareCompany(company)
        }

    }
    
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true
    }

    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            let company = fetchedResultsController?.object(at: indexPath) as? QSCompany
            if let aCompany = company {
                managedObjectContext?.delete(aCompany)
            }
            try? managedObjectContext?.save()
        }
    }


//pragma mark - Actions
  
    @IBAction func insertNewCompany(_ sender: Any) {
        let alertController = UIAlertController(title: "New company", message: nil, preferredStyle: .alert)
        alertController.addTextField(configurationHandler: { textField in
            textField.placeholder = "Enter company's name"
        })
        alertController.addAction(UIAlertAction(title: "Add", style: .default, handler: { action in
            self.createCompanyWithName(name: (alertController.textFields?.first?.text)!)
        }))
        present(alertController, animated: true)
    }


    @IBAction func didTapSynchronize(_ sender: Any) {
        synchronize(withCompletion: nil)
    }

    
    func synchronize(withCompletion completion: ((_ error: Error?) -> Void)? = nil) {
        showLoading(true)
        print("synchronize called")
        self.synchronizer?.synchronize(completion: { error in
            self.showLoading(false)
            if error != nil {
                var alertController: UIAlertController? = nil
                if let anError = error {
                    print("Sync Error : \(anError)")
                    alertController = UIAlertController(title: "Sync Error", message: "Error: \(anError)", preferredStyle: .alert)
                }
                alertController?.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
                if let aController = alertController {
                    self.present(aController, animated: true)
                }
            } else {
                let zoneID: CKRecordZoneID? = self.synchronizer?.modelAdapters().first?.recordZoneID()
                if zoneID != nil {
                    self.synchronizer?.subscribeForChanges(in: zoneID!) { error in
                        if error != nil {
                            if let anError = error {
                                print("Failed to subscribe with error: \(anError)")
                            }
                        } else {
                            print("Subscribed for notifications")
                        }
                    }
                }
            }
            if (completion != nil) {
                completion!(error)
            }
            
        })
    }
    

    func shareCompany(_ company: QSCompany) {
        sharingCompany = company
        synchronize(withCompletion: { error in
            if error == nil {
                var sharingController: UICloudSharingController?
                let share: CKShare? = self.synchronizer?.share(for: company)
                let container = CKContainer(identifier: (self.synchronizer?.containerIdentifier)! )
                if share != nil {
                    if let aShare = share {
                        sharingController = UICloudSharingController(share: aShare, container: container)
                    }
                } else {
                    sharingController = UICloudSharingController(preparationHandler: { controller, preparationCompletionHandler in
                        self.synchronizer?.share(object: company, publicPermission: CKShareParticipantPermission.readOnly, participants: []) { share, error in
                            if let aName = company.name {
                                share?[CKShareTitleKey] = aName as CKRecordValue
                            }
                            preparationCompletionHandler(share, container, error)
                        }
                    })
                }
                sharingController?.availablePermissions = [.allowPublic, .allowReadOnly, .allowReadWrite]
                sharingController?.delegate = self
                if let aController = sharingController {
                    self.present(aController, animated: true)
                }
            }
        })
    }

    
    
// pragma mark - UICloudSharingControllerDelegate
    
    func itemTitle(for csc: UICloudSharingController) -> String? {
        return self.sharingCompany.name
    }

    func cloudSharingController(_ csc: UICloudSharingController, failedToSaveShareWithError error: Error) {
        print("Error saving: \(error)")
    }

    func cloudSharingControllerDidSaveShare(_ csc: UICloudSharingController) {
        synchronizer?.saveShare(csc.share!, for: sharingCompany)
        tableView.reloadData()
    }

    func cloudSharingControllerDidStopSharing(_ csc: UICloudSharingController) {
        synchronizer?.deleteShare(for: sharingCompany)
        tableView.reloadData()
    }

    
//pragma mark - Loading
    
    func showLoading(_ loading: Bool) {
        if loading {
            syncButton.isHidden = true
            loadingView?.isHidden = false
            indicatorView.startAnimating()
        } else {
            syncButton.isHidden = false
            loadingView?.isHidden = true
            indicatorView.stopAnimating()
        }
    }
    

    
}
